### PR TITLE
Allow user to specify the init values for each component of the tuple

### PR DESCRIPTION
**Pull request by imikejackson**
Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

### CODE REVIEW

File: Source/SIMPLib/CoreFilters/CreateDataArray.cpp [hunk 1]
```diff
@@ -56,89 +56,97 @@
...
+      {                                                                                                                                                                                                \
+        return 0;                                                                                                                                                                                      \
+      }                                                                                                                                                                                                \
+      return value;                                                                                                                                                                                    \
+    }                                                                                                                                                                                                  \
+  };
+
+CDA_CREATE_CONVERTOR(uint8_t, toUInt)
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/CreateDataArray.cpp:82]: All the non class functions can be put into an anonymous namespace.

File: Source/SIMPLib/CoreFilters/CreateDataArray.cpp [hunk 2]
```diff
@@ -56,89 +56,97 @@
...
+CDA_CREATE_CONVERTOR(uint64_t, toULongLong)
+CDA_CREATE_CONVERTOR(int64_t, toLongLong)
+CDA_CREATE_CONVERTOR(bool, toInt)
+
+CDA_CREATE_CONVERTOR(float, toDouble)
+CDA_CREATE_CONVERTOR(double, toDouble)
+
+#define CDA_SEP ";"
+#define CDA_LB "["
+#define CDA_RB "]"
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/CreateDataArray.cpp]:
```suggestion
constexpr char CDA_SEP = ';';
constexpr char CDA_LB = '[';
constexpr char CDA_RB = ']';
```
Change to constexpr char variables.

File: Source/SIMPLib/CoreFilters/CreateDataArray.cpp [hunk 3]
```diff
@@ -56,89 +56,97 @@
...
+      {                                                                                                                                                                                                \
+        return 0;                                                                                                                                                                                      \
+      }                                                                                                                                                                                                \
+      return value;                                                                                                                                                                                    \
+    }                                                                                                                                                                                                  \
+  };
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/CreateDataArray.cpp]:
```
#define CDA_CREATE_CONVERTOR(type, func)                                                                                                                                                               \
  template <>                                                                                                                                                                                          \
  struct convert_from<type>                                                                                                                                                                            \
  {                                                                                                                                                                                                    \
    static type convert(const QString& input)                                                                                                                                                          \
    {                                                                                                                                                                                                  \
      bool ok = false;                                                                                                                                                                                 \
      type value = static_cast<type>(input.func(&ok));                                                                                                                                                    \
      if(!ok)                                                                                                                                                                                          \
      {                                                                                                                                                                                                \
        return 0;                                                                                                                                                                                      \
      }                                                                                                                                                                                                \
      return value;                                                                                                                                                                                    \
    }                                                                                                                                                                                                  \
  };

template <class T>
struct convert_from
{
};

CDA_CREATE_CONVERTOR(uint8_t, toUInt)
CDA_CREATE_CONVERTOR(int8_t, toInt)
CDA_CREATE_CONVERTOR(uint16_t, toUShort)
CDA_CREATE_CONVERTOR(int16_t, toShort)
CDA_CREATE_CONVERTOR(uint32_t, toUInt)
CDA_CREATE_CONVERTOR(int32_t, toInt)
CDA_CREATE_CONVERTOR(uint64_t, toULongLong)
CDA_CREATE_CONVERTOR(int64_t, toLongLong)
CDA_CREATE_CONVERTOR(bool, toInt)
```

Using template specializations in the macro removes the need for the conversion template parameter in createInitVector(). The convert function can also be made static.

File: Source/SIMPLib/CoreFilters/CreateDataArray.cpp [hunk 4]
```diff
@@ -429,71 +439,139 @@ void CreateDataArray::execute()
...
+    initializeArrayWithBool(m_OutputArrayPtr.lock(), m_InitializationType, values);
   }
 }
 
 // -----------------------------------------------------------------------------
 //
 // -----------------------------------------------------------------------------
-template <typename T> void CreateDataArray::checkInitialization(QString dataArrayName)
+template <typename T>
+void CreateDataArray::checkInitialization(QString dataArrayName)
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/CreateDataArray.cpp:524]: Putting in a list of components fails for float and double.

File: Source/SIMPLib/CoreFilters/CreateDataArray.cpp [hunk 5]
```diff
@@ -429,71 +439,139 @@ void CreateDataArray::execute()
...
 {
   if(m_InitializationType == Manual)
   {
     bool ok;
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/CreateDataArray.cpp]: Should be initialized.

File: Source/SIMPLib/CoreFilters/CreateDataArray.cpp [hunk 6]
```diff
@@ -429,71 +439,139 @@ void CreateDataArray::execute()
...
+      QStringList tokens = input.split(CDA_SEP, QString::SplitBehavior::KeepEmptyParts);
+      if(m_NumberOfComponents != tokens.size() && tokens.size() != 1)
+      {
+        QString ss = QString("Number of initialization components does not match number of components.%1 vs %2").arg(tokens.size()).arg(m_NumberOfComponents);
+        setErrorCondition(-5459, ss);
+        return;
+      }
+      for(const auto& token : tokens)
+      {
+        std::ignore = token.toDouble(&ok);
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/CreateDataArray.cpp]: We should check the conversion with the actual type. For example if you enter -1 for an unsigned type, it the Qt functions fail to convert it resulting in 0, but it is a silent error during dataCheck().

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Comment by imikejackson** [comment]
All the requested changes have been made.

File: Source/SIMPLib/CoreFilters/CreateDataArray.cpp [hunk 7]
```diff
@@ -110,27 +117,58 @@ std::vector<T> createInitVector(const QString& initializationValue, int32_t numC
     {
       for(int32_t i = 0; i < numComponents; i++)
       {
-        values[i] = convert.convert(tokens[0]);
+        conversionResult = convert.convert(tokens[0]);
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/CreateDataArray.cpp]:
```suggestion
        conversionResult = convert_from<T>::convert(tokens[0]);
```
Call statically here and in the other spots. Can remove local variables afterwards.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Pull request merged**